File: api/src/routes/enroll.ts
import type { FastifyInstance } from "fastify";
import { generateEmbedding } from "../services/faceService.js";
import { userStore } from "../store.js";
import type { EnrollRequest, EnrollResponse, ErrorResponse } from "../types.js";

export async function enrollRoutes(app: FastifyInstance): Promise<void> {
    app.post<{ Body: EnrollRequest; Reply: EnrollResponse | ErrorResponse }>(
        "/api/enroll",
        async (request, reply) => {
            const { userId, name, image } = request.body;

            if (!userId || !name || !image) {
                return reply.status(400).send({
                    success: false,
                    error: "Missing required fields: userId, name, image",
                });
            }

            if (userStore.exists(userId)) {
                return reply.status(409).send({
                    success: false,
                    error: `User "${userId}" is already enrolled`,
                });
            }

            try {
                const { descriptor, livenessScore } = await generateEmbedding(image);

                if (livenessScore < 0.5) {
                    return reply.status(422).send({
                        success: false,
                        error: "Liveness check failed. Please use a live camera capture.",
                    });
                }

                userStore.add({
                    id: userId,
                    name,
                    descriptor: Array.from(descriptor),
                    enrolledAt: new Date().toISOString(),
                });

                return reply.status(201).send({
                    success: true,
                    userId,
                    message: `User "${name}" enrolled successfully`,
                });
            } catch (err) {
                const message = err instanceof Error ? err.message : "Unknown error";

                if (message === "NO_FACE_DETECTED") {
                    return reply.status(422).send({
                        success: false,
                        error: "No face detected in the image",
                    });
                }
                if (message === "MULTIPLE_FACES_DETECTED") {
                    return reply.status(422).send({
                        success: false,
                        error: "Multiple faces detected. Please provide an image with exactly one face.",
                    });
                }

                request.log.error(err, "Enrollment failed");
                return reply.status(500).send({
                    success: false,
                    error: "Internal server error during enrollment",
                });
            }
        }
    );
}

File: api-go/internal/handlers/enroll.go
package handlers

import (
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"time"

	faceservice "facial-auth/internal/face"
	"facial-auth/internal/store"
	"facial-auth/internal/types"
)

// EnrollHandler handles POST /api/enroll.
type EnrollHandler struct {
	faceService *faceservice.Service
	store       *store.UserStore
}

// NewEnrollHandler creates an EnrollHandler.
func NewEnrollHandler(fs *faceservice.Service, s *store.UserStore) *EnrollHandler {
	return &EnrollHandler{faceService: fs, store: s}
}

// ServeHTTP handles POST /api/enroll.
func (h *EnrollHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	var req types.EnrollRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeJSON(w, http.StatusBadRequest, types.ErrorResponse{
			Success: false,
			Error:   "Corpo da requisição inválido",
		})
		return
	}

	if req.UserID == "" || req.Name == "" || req.Image == "" {
		writeJSON(w, http.StatusBadRequest, types.ErrorResponse{
			Success: false,
			Error:   "Campos obrigatórios não informados: userId, name, image",
		})
		return
	}

	if h.store.Exists(req.UserID) {
		writeJSON(w, http.StatusConflict, types.ErrorResponse{
			Success: false,
			Error:   fmt.Sprintf("O usuário %q já está cadastrado", req.UserID),
		})
		return
	}

	result, err := h.faceService.GenerateEmbedding(req.Image)
	if err != nil {
		switch {
		case errors.Is(err, faceservice.ErrNoFaceDetected):
			writeJSON(w, http.StatusUnprocessableEntity, types.ErrorResponse{
				Success: false,
				Error:   "Nenhum rosto detectado na imagem",
			})
		case errors.Is(err, faceservice.ErrMultipleFacesDetected):
			writeJSON(w, http.StatusUnprocessableEntity, types.ErrorResponse{
				Success: false,
				Error:   "Múltiplos rostos detectados. Envie uma imagem com apenas um rosto.",
			})
		default:
			writeJSON(w, http.StatusInternalServerError, types.ErrorResponse{
				Success: false,
				Error:   "Erro interno do servidor ao realizar o cadastro",
			})
		}
		return
	}

	if result.LivenessScore < 0.5 {
		writeJSON(w, http.StatusUnprocessableEntity, types.ErrorResponse{
			Success: false,
			Error:   "Verificação de vivacidade falhou. Utilize uma captura ao vivo da câmera.",
		})
		return
	}

	h.store.Add(&types.User{
		ID:         req.UserID,
		Name:       req.Name,
		Descriptor: faceservice.DescriptorToSlice(result.Descriptor),
		EnrolledAt: time.Now().UTC().Format(time.RFC3339),
		Status:     types.StatusAtivo,
	})

	writeJSON(w, http.StatusCreated, types.EnrollResponse{
		Success: true,
		UserID:  req.UserID,
		Message: fmt.Sprintf("Usuário %q cadastrado com sucesso", req.Name),
	})
}
